sw/airborne/firmwares/rotorcraft/guidance/guidance_h_ref.c

```diff
@@ -42,229 +40,156 @@ static const int32_t gh_max_accel = BFP_OF_REAL(GUIDANCE_H_REF_MAX_ACCEL, GH_ACC
-  }
-  /* restore gh_ref.speed range (Q14.17) */
-  INT32_VECT2_LSHIFT(gh_ref.max_vel, gh_ref.max_vel, (GH_SPEED_REF_FRAC - GH_MAX_SPEED_REF_FRAC));
+  // copy accel
+  VECT2_COPY(gh_ref.accel, accel_sp);
 }
 
-static void gh_compute_ref_max_accel(struct Int32Vect2 *ref_vector)
+void gh_update_ref_from_speed_sp(struct FloatVect2 speed_sp)
 {
-  /* Bound ref to max accel along reference vector.
-   * If angle can't be computed, simply set both axes to max magnitude/sqrt(2).
-   */
-  if (ref_vector->x == 0 && ref_vector->y == 0) {
-    gh_ref.max_accel.x = gh_ref.max_accel.y = gh_max_accel * 0.707;
-  } else {
-    gh_compute_route_ref(ref_vector);
-    /* Compute maximum acceleration*/
-    gh_ref.max_accel.x = INT_MULT_RSHIFT(gh_max_accel, gh_ref.c_route_ref, INT32_TRIG_FRAC);
-    gh_ref.max_accel.y = INT_MULT_RSHIFT(gh_max_accel, gh_ref.s_route_ref, INT32_TRIG_FRAC);
-  }
-}
+  struct FloatVect2 pos_step, speed_step;
 
-static void gh_compute_ref_max_speed(struct Int32Vect2 *ref_vector)
-{
-  /* Bound ref to max speed along reference vector.
-   * If angle can't be computed, simply set both axes to max magnitude/sqrt(2).
-   */
-  if (ref_vector->x == 0 && ref_vector->y == 0) {
-    gh_ref.max_vel.x = gh_ref.max_vel.y = gh_ref.max_speed_int * 0.707;
-  } else {
-    gh_compute_route_ref(ref_vector);
-    /* Compute maximum reference x/y velocity from absolute max_speed */
-    gh_ref.max_vel.x = INT_MULT_RSHIFT(gh_ref.max_speed_int, gh_ref.c_route_ref, INT32_TRIG_FRAC);
-    gh_ref.max_vel.y = INT_MULT_RSHIFT(gh_ref.max_speed_int, gh_ref.s_route_ref, INT32_TRIG_FRAC);
-  }
-  /* restore gh_ref.speed range (Q14.17) */
-  INT32_VECT2_LSHIFT(gh_ref.max_vel, gh_ref.max_vel, (GH_SPEED_REF_FRAC - GH_MAX_SPEED_REF_FRAC));
+  VECT2_SMUL(pos_step, gh_ref.speed, gh_ref.dt);
+  VECT2_SMUL(speed_step, gh_ref.accel, gh_ref.dt);
+
+  struct Int64Vect2 pos_update;
+  pos_update.x = BFP_OF_REAL(pos_step.x, GH_POS_REF_FRAC);
+  pos_update.y = BFP_OF_REAL(pos_step.y, GH_POS_REF_FRAC);
+
+  VECT2_ADD(gh_ref.pos, pos_update);
+  VECT2_ADD(gh_ref.speed, speed_step);
+
+  // compute speed error
+  struct FloatVect2 speed_err;
+  VECT2_DIFF(speed_err, gh_ref.speed, speed_sp);
+  // compute accel from speed_sp
+  struct FloatVect2 accel_sp;
+  VECT2_SMUL(accel_sp, speed_err, -gh_ref.inv_tau);
+
+  gh_saturate_accel(&accel_sp);
+
+  // copy accel
+  VECT2_COPY(gh_ref.accel, accel_sp);
 }
 
-/** saturate reference accelerations */
-static void gh_saturate_ref_accel(void)
+static void gh_saturate_speed(struct FloatVect2 *speed_sp)
 {
-  /* Saturate accelerations */
-  BoundAbs(gh_ref.accel.x, gh_ref.max_accel.x);
-  BoundAbs(gh_ref.accel.y, gh_ref.max_accel.y);
+  // Speed squared
+  float v_norm2 = VECT2_NORM2(*speed_sp);
+
+  // Apply saturation if above max speed
+  if (v_norm2 > (gh_ref.max_speed * gh_ref.max_speed)) {
+    // speed_sp/sqrt(v_norm2)*vmax
+    float factor = gh_ref.max_speed / sqrtf(v_norm2);
+    VECT2_SMUL(*speed_sp, *speed_sp, factor);
+  }
```

Comment:
if speed is saturated, shouldn't the acceleration be saturated as well (in the "direction" of speed saturation) ?

Reply:
I don't think so, why would that be the case? The 'desired speed' (vel_sp) may be very large for a waypoint far away, but the saturated vel_sp will have a realistic value. From that the acceleration setpoint is calculated, which may or may not be saturated, Depending on the current speed of the reference model, the desired acceleration vector may have a different direction than the desired velocity vector. The acceleration should be saturated such that its direction is preserved. I hope I understood your question correctly though 😄